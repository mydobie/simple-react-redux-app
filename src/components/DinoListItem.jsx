import React from 'react';
import PropTypes from 'prop-types';
import { Form } from 'react-bootstrap';

const DinoListItem = (props) => {
  const { dinoName, dinoId, checked, changeCheckBox } = props;
  return (
    <li>
      {/* EXAMPLE: Checkbox with label */}
      <Form.Check
        onClick={(e) => {
          changeCheckBox(dinoId, e.target.checked);
        }}
        id={`cb_${dinoId}`}
        defaultChecked={checked}
        label={dinoName}
      />
    </li>
  );
};

DinoListItem.propTypes = {
  dinoName: PropTypes.string,
  dinoId: PropTypes.string.isRequired,
  checked: PropTypes.bool,
  changeCheckBox: PropTypes.func,
};
DinoListItem.defaultProps = {
  dinoName: '',
  checked: false,
  changeCheckBox: () => {},
};

export default DinoListItem;
