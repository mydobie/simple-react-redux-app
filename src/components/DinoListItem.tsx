import React, { ReactElement } from 'react';
import { Form } from 'react-bootstrap';

// ** Main component type */
type DinoListItemProps = {
  dinoName: string;
  dinoId: string;
  checked: boolean;
  // EXAMPLE: Passing a function as a prop (so data bubbles to parent)
  changeCheckBox: (dinoId: string, isChecked: boolean) => void;
  testid?: string;
};

// *** Main component ***
const DinoListItem = ({
  dinoName,
  dinoId,
  checked,
  changeCheckBox,
  testid,
}: DinoListItemProps): ReactElement => (
  <li data-testid={testid}>
    {/* EXAMPLE: Checkbox with label */}
    <Form.Check
      onClick={(e) => {
        changeCheckBox(dinoId, (e.target as HTMLInputElement).checked);
      }}
      id={`cb_${dinoId}`}
      defaultChecked={checked}
      label={dinoName}
    />
  </li>
);

export default DinoListItem;
