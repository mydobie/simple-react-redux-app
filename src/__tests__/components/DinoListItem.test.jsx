/* eslint-disable react/react-in-jsx-scope */
import { shallow } from 'enzyme';
import { axe } from 'jest-axe';
import DinoListItem from '../../components/DinoListItem';

// EXAMPLE: Mocking a function sent in as a prop
const mockChangeCheckbox = jest.fn();

describe('Loading icon tests', () => {
  let wrapper = '';
  afterEach(() => {
    mockChangeCheckbox.mockClear();
  });

  beforeEach(() => {
    wrapper = shallow(
      <DinoListItem
        dinoName='tRex'
        dinoId='4'
        checked={false}
        changeCheckBox={mockChangeCheckbox}
      />
    );
    wrapper.update();
  });

  test('Is accessible', async () => {
    const results = await axe(`<main><ul>${wrapper.html()}</ul></main>`); // NOTE ul is required to prevent landmark error
    expect(results).toHaveNoViolations();
  });
  test('When checkbox is clicked, changeCheckBox is called', () => {
    // EXAMPLE: Checking for a React Bootstrap element.
    // In the component it is listed as Form.Check
    // If you are unsure what an element is really called, then look at the snapshot
    expect(wrapper).toMatchSnapshot();
    wrapper
      .find('FormCheck')
      .first()
      .simulate('click', { target: { checked: true } });
    expect(mockChangeCheckbox).toHaveBeenCalledTimes(1);
    expect(mockChangeCheckbox.mock.calls[0][0]).toBe('4');
    expect(mockChangeCheckbox.mock.calls[0][1]).toBe(true); // second attribute sent
  });

  test('When checkbox is unchecked, changeCheckBox is called', () => {
    // EXAMPLE: Simulate a checkbox click
    wrapper
      .find('FormCheck')
      .first()
      .simulate('click', { target: { checked: false } });
    expect(mockChangeCheckbox).toHaveBeenCalledTimes(1);
    expect(mockChangeCheckbox.mock.calls[0][0]).toBe('4');
    expect(mockChangeCheckbox.mock.calls[0][1]).toBe(false); // second attribute sent
  });
});
