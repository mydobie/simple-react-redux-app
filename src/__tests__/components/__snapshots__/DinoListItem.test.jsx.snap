// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Loading icon tests When checkbox is clicked, changeCheckBox is called 1`] = `
<li>
  <FormCheck
    defaultChecked={false}
    id="cb_4"
    label="tRex"
    onClick={[Function]}
  />
</li>
`;
